share/shwap/p2p/shrex/middleware.go

```diff
@@ -1,41 +1,50 @@
 package shrex
 
 import (
+	"context"
 	"sync/atomic"
 
-	logging "github.com/ipfs/go-log/v2"
 	"github.com/libp2p/go-libp2p/core/network"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
 )
 
-var log = logging.Logger("shrex/middleware")
-
 type Middleware struct {
 	// concurrencyLimit is the maximum number of requests that can be processed at once.
 	concurrencyLimit int64
 	// parallelRequests is the number of requests currently being processed.
 	parallelRequests atomic.Int64
-	// numRateLimited is the number of requests that were rate limited.
-	numRateLimited atomic.Int64
+
+	rateLimiterCounter metric.Int64Counter
 }
 
-func NewMiddleware(concurrencyLimit int) *Middleware {
-	return &Middleware{
-		concurrencyLimit: int64(concurrencyLimit),
+func NewMiddleware(concurrencyLimit int) (*Middleware, error) {
+	rateLimiter, err := meter.Int64Counter("shrex_rate_limit_counter",
+		metric.WithDescription("concurrency limit of the shrex server"),
```

Comment:
This needs to be set with WithMetrics, otherwise it can be initied before metrics collector and will send all data to dev/null

Reply:
added to metrics